PenLineModel.cpp

```diff
@@ -5,10 +5,39 @@
 //Our includes
 #include "PenLineModel.h"
 
-class PenLineModelUndoCommand : public QUndoCommand {
+class PenLineModelAddLineCommand : public QUndoCommand {
 public:
-    PenLineModelUndoCommand(PenLineModel *model, const QVector<PenLine> &oldLines, const QVector<PenLine> &newLines) :
-        QUndoCommand("Draw Line", nullptr), m_penLineModel(model), m_oldLines(oldLines), m_newLines(newLines) {}
+    PenLineModelAddLineCommand(PenLineModel *model, PenLine addedLine) :
+        QUndoCommand("Draw Line", nullptr), m_penLineModel(model), m_addedLine(std::move(addedLine)) {}
+
+    void undo() override
+    {
+        Q_ASSERT(!m_penLineModel->m_lines.isEmpty());
+        size_t index = m_penLineModel->m_lines.size() - 1;
+        qDebug() << "PenLineModelAddLineCommand::undo " << index << " " << index;
+        m_penLineModel->beginRemoveRows(QModelIndex(), index, index);
+        m_penLineModel->m_lines.removeLast();
+        m_penLineModel->endRemoveRows();
+    }
+
+    void redo() override
+    {
+        size_t index = m_penLineModel->m_lines.size();
+        qDebug() << "PenLineModelAddLineCommand::redo " << index << " " << index;
+        m_penLineModel->beginInsertRows(QModelIndex(),index, index);
+        m_penLineModel->m_lines.append(m_addedLine);
+        m_penLineModel->endInsertRows();
+    }
+
+private:
+    PenLineModel *m_penLineModel;  // not owned
```

Comment:
I didn't catch this is the last code review, but it's a good idea to make these QPointer<PenLineModel*>. A QPointer is a special QObject weak pointer that will go null if PenLineModel is delete somehow. Of course it's a memory bug if that happens, but it's easier to see what's going on.